[#624] Upgrade dependencies

roberlander2 opened the pull request:
## Description
This PR upgrades `go.mod` packages to their latest versions (except github.com/getkin/kin-openapi, which may have an issue validating example requests in the latest version).

**Resolves #624**
**Resolves #593**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [ ] Within a week
- [x] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My change requires updating the documentation.
- [x] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

shurwit commented on driver/web/auth.go:
```diff
@@ -15,450 +15,195 @@
 
-func newServicesAuth(coreAPIs *core.APIs, serviceRegManager *authservice.ServiceRegManager, serviceID string, logger *logs.Logger) (*ServicesAuth, error) {
+func newServicesAuth(serviceRegManager *authservice.ServiceRegManager, serviceID string) (*tokenauth.StandardHandler, error) {
 	servicesScopeAuth := authorization.NewCasbinScopeAuthorization("driver/web/authorization_services_policy.csv", serviceID)
```
It looks like we are passing the same file as both a scope and permission policy. This issue was not introduced here but we should fix it because it may result in unexpected behavior.

shurwit commented on driver/web/auth.go:
```diff
@@ -15,450 +15,195 @@
-	}
+	check := func(claims *tokenauth.Claims, req *http.Request) (int, error) {
+		if claims.Admin {
+			return http.StatusUnauthorized, errors.ErrorData(logutils.StatusInvalid, "admin claim", nil)
```
I think we probably want to return `StatusForbidden` in all of these cases to avoid triggering unnecessary token refresh attempts. It is also more accurate since `StatusUnauthorized` should be used for missing/invalid credentials while `StatusForbidden` should be used for valid credentials that do not grant access to the requested resource.

shurwit commented on driver/web/auth.go:
```diff
@@ -15,450 +15,195 @@
+// TPsAuth
 
-func newTPsAuth(coreAPIs *core.APIs, serviceRegManager *authservice.ServiceRegManager, logger *logs.Logger) (*TPsAuth, error) {
+func newTPsAuth(serviceRegManager *authservice.ServiceRegManager) (*tokenauth.StandardHandler, error) {
```
The 's' in `TPs` should be capitalized (ie `TPS`)

shurwit commented on driver/web/docs/schemas/auth/ServiceAccount.yaml:
```diff
@@ -1,11 +1,6 @@
 required:
-  - account_id
```
Shouldn't `account_id` be `required`, `readyOnly`?

roberlander2 replied:
No, because `account_id` is expected in the request body when creating a service account by copying an existing one.

shurwit reviewed (commented):
Thanks @roberlander2, this looks good! Just a few minor comments below.

shurwit reviewed (approved):
Thanks @roberlander2, LGTM! @petyos please take a look when you get a chance. There are lots of changes here but most of them are minor logging fixes/renames. The important part to focus on is probably the changes to the docs. Please also take a look at my changes to the makefile and let me know if you see any issues. You can refer to my comment here for information about `govulncheck`: https://github.com/rokwire/groups-building-block/issues/319#issuecomment-1404126468. Thanks!

petyos reviewed (commented):
Hi @roberlander2  @shurwit , thanks for this PR. It looks good. I have two comments.

1. When I run it it fails with this error on startup:
`time="2023-01-30T09:32:32+02:00" level=fatal msg="error on openapi3 validate - invalid paths: invalid path /admin/auth/login: invalid operation POST: example email: Error at \"/device/id\": property \"id\" is missing\nSchema:\n  {\n    \"properties\": {\n      \"device_id\": {\n        \"type\": \"string\"\n      },\n      \"id\": {\n        \"readOnly\": true,\n        \"type\": \"string\"\n      },\n      \"os\": {\n        \"type\": \"string\"\n      },\n      \"type\": {\n        \"enum\": [\n          \"mobile\",\n          \"web\",\n          \"desktop\",\n          \"other\"\n        ],\n        \"type\": \"string\"\n      }\n    },\n    \"required\": [\n      \"id\",\n      \"type\"\n    ],\n    \"type\": \"object\"\n  }\n\nValue:\n  {\n    \"device_id\": \"5555\",\n    \"os\": \"Android\",\n    \"type\": \"mobile\"\n  }\n" service_name=core`

2. I am wondering why you have removed [entity]Fields.yaml files. As general in the schemas we could differentiate them into 3 categories:
- Model schemas
The reader understands the data model of Core BB. Every entity contains its fields (id, name, etc) and also the composition fields pointing to other entities. The plain fields are also separated in a file with ...Fields.yaml postfix. This file is also used in point 3 for reusing these files.

- APIs Requests schemas

- APIs Responses schemas

//////

For example let's look the response of this API before your changes:

https://api-dev.rokwire.illinois.edu/core/doc/ui/index.html#/Admin/get_admin_application_login_sessions

```
[
  {
    "id": "string",
    "auth_type_code": "string",
    "app_type_id": "string",
    "app_type_identifier": "string",
    "anonymous": true,
    "identifier": "string",
    "account_auth_type_id": "string",
    "account_auth_type_identifier": "string",
    "device_id": "string",
    "ip_address": "string",
    "refresh_tokens_count": 0,
    "state": "string",
    "state_expires": "string",
    "mfa_attempts": 0,
    "date_refreshed": "string",
    "date_updated": "string",
    "date_created": "string"
  }
]
```
After your changes this API response has been changed and additional no needed structures are added in the response:

```
[
  {
    "id": "string",
    "anonymous": true,
    "identifier": "string",
    "app_org": {
      "id": "string",
      "app_id": "string",
      "org_id": "string",
      "services_ids": [
        "string"
      ],
      "identity_provider_settings": [
        {
          "identity_provider_id": "string",
          "user_identifier_field": "string",
          "external_id_fields": {
            "additionalProp1": "string",
            "additionalProp2": "string",
            "additionalProp3": "string"
          },
          "first_name_field": "string",
          "middle_name_field": "string",
          "last_name_field": "string",
          "email_field": "string",
          "roles_field": "string",
          "groups_field": "string",
          "user_specific_fields": [
            "string"
          ],
          "roles": {
            "additionalProp1": "string",
            "additionalProp2": "string",
            "additionalProp3": "string"
          },
          "groups": {
            "additionalProp1": "string",
            "additionalProp2": "string",
            "additionalProp3": "string"
          },
          "always_sync_profile": true
        }
      ],
      "supported_auth_types": [
        {
          "app_type_id": "string",
          "supported_auth_types": [
            {
              "auth_type_id": "string",
              "params": {
                "additionalProp1": {}
              }
            }
          ]
        }
      ],
      "login_session_settings": {
        "max_concurrent_sessions": 0,
        "inactivity_expire_policy": {
          "active": true,
          "inactivity_period": 0
        },
        "time_since_login_expire_policy": {
          "active": true,
          "time_since_login_period": 0
        },
        "yearly_expire_policy": {
          "active": true,
          "day": 0,
          "month": 0,
          "hour": 0,
          "min": 0
        }
      }
    },
    "app_type": {
      "id": "string",
      "identifier": "string",
      "name": "string",
      "versions": [
        "string"
      ]
    },
    "account_auth_type": {
      "id": "string",
      "code": "string",
      "identifier": "string",
      "params": {
        "additionalProp1": {}
      },
      "active": true,
      "unverified": true,
      "credential": {
        "id": "string",
        "value": {}
      }
    },
    "device_id": "string",
    "ip_address": "string",
    "refresh_tokens_count": 0,
    "state": "string",
    "state_expires": "string",
    "mfa_attempts": 0,
    "date_refreshed": "string",
    "date_updated": "string",
    "date_created": "string"
  }
]
```
As general if we want to remove the ..Fields.yaml files then we need to replicate these files in the "APIs Responses schemas" where they are used. This is why we introduced them in the "Model schemas" so that we can reuse them in the APIs responses.

Let me know if you have any comments. Thanks!

roberlander2 commented:
Thanks @petyos,

1. Thanks for catching this. I did not see that the `kin-openapi` package was updated to `v0.113.0`. This version changes the definition of `readOnly` from the definition that we have been operating under to this point. `v0.110.0` seems to me to be the most recent version that does not change this definition, so downgrading the version in `go.mod` should allow you to run successfully. I am not sure whether we want to upgrade to `v0.113.0`, but if so, we will need to update our schemas to fit the new `readOnly` definition.

2. I think I understand what you are saying about the `..Fields.yaml` schemas, and I agree with your comment about splitting our schemas into 3 categories. The reason why I removed the existing `..Fields.yaml` schemas is because it seemed to me to be straightforward to merge them into the standard model schemas given the request and response schemas expected by the current version. You are correct about the modified `LoginSession` schema though, and I will go back and make sure it matches the response sent by the current version of the Core BB. I will also check to make sure the other schemas I changed match the behavior of the current version of the Core BB.

I think it is fine if we need to add new model schemas for new APIs that are slightly different from existing model schemas. This would be similar to `PartialAccount.yaml`, which is an account model, but slightly different from `Account.yaml`.

Please let me know what you think about this. Thanks!

petyos commented:
Hi @roberlander2 ,

1. 
I think for now downgrading to v0.110.0 should be good.

_I think it is fine if we need to add new model schemas for new APIs that are slightly different from existing model schemas. This would be similar to PartialAccount.yaml, which is an account model, but slightly different from Account.yaml._

Basically we already have all needed types of schemas - `Model schemas`,` APIs requests schemas` and `APIs responses schemas`.

- Model schemas
These schemas are a mirror of the structs which are placed in the model package. The reader can understand the model without reading Golang code but directly in the documentation.These schemas are not related with APIs. The only relation is that we reuse the ...Fields.yaml files for responses but this is just for files reusing as we mentioned.
If the reader wants to call the Core BB APIs he should look at the schemas in the following two sections(requests and responses) below but not the model schemas whose purpose is to explain the data model only.

- APIs Requests and Responses schemas
For very simple data models the model schemas and the requests and responses schemas will be the same but for more complex models which is our case in Core BB they differ. Various requests and responses could have a substructure from the main model as we have in our APIs.

### Regarding PartialAccount.yaml 
If we want to be correct the place for PartialAccount.yaml is not in the model schemas but it should be in the responses schemas section. The reason is that in the model package in Golang we do not have a PartialAccount struct. This entity is needed for some of our APIs as a response. I do not say that we need to move this now in the response but I just want to be sure you understand what I mean. Imagine that we need another API which requires us to expose another substructure of our Account struct - PartialAccount2 etc.

Please let me know if you ahve any questions or comments. Thanks!

shurwit commented:
Hi @roberlander2 and @petyos, thanks for the discussion and feedback!
> 1.  I think for now downgrading to v0.110.0 should be good.
I think we need to find a way to use the latest version of the package with the new definition of `readOnly` or else we will not be able to stay up to date with the latest changes going forward. My opinion is that it is better to take care of this now while making related changes than to put it off and have to deal with it again later.

> Basically we already have all needed types of schemas - `Model schemas`,` APIs requests schemas` and `APIs responses schemas`.
> 
> * Model schemas
>   These schemas are a mirror of the structs which are placed in the model package. The reader can understand the model without reading Golang code but directly in the documentation.These schemas are not related with APIs. The only relation is that we reuse the ...Fields.yaml files for responses but this is just for files reusing as we mentioned.
>   If the reader wants to call the Core BB APIs he should look at the schemas in the following two sections(requests and responses) below but not the model schemas whose purpose is to explain the data model only.
> * APIs Requests and Responses schemas
>   For very simple data models the model schemas and the requests and responses schemas will be the same but for more complex models which is our case in Core BB they differ. Various requests and responses could have a substructure from the main model as we have in our APIs.
> 
> ### Regarding PartialAccount.yaml
> If we want to be correct the place for PartialAccount.yaml is not in the model schemas but it should be in the responses schemas section. The reason is that in the model package in Golang we do not have a PartialAccount struct. This entity is needed for some of our APIs as a response. I do not say that we need to move this now in the response but I just want to be sure you understand what I mean. Imagine that we need another API which requires us to expose another substructure of our Account struct - PartialAccount2 etc.

Thanks @petyos for the detailed explanation of your view of the organization for the schemas! This is very clear and makes it easy for us to understand where the disconnect is. We have been thinking about the schemas differently from how you described them. Here is our understanding and some explanation of the reasoning for it:

### Schemas
* **Model schemas:** These schemas reflect the standard representation of a model in our system from the perspective of the API caller. They may or may not match the Go model structs in our model package exactly depending on whether all fields we store internally are exposed through the APIs. These should be reusable components that are used directly throughout the docs to represent this data type in the requests/responses to APIs. 
* **APIs Requests and Responses schemas:** If the schema of the data to be sent in an API request/response differs from our existing models we can create a schema specific to this API to reflect this change. This should not be as common as using the model schemas directly in the request/response, and these should never be an exact duplicate of a model schema. 

### Regarding PartialAccount.yaml
While I believe the naming of this type is not ideal, this was meant to reflect a standard schema that can be sent through any API that should only expose the minimum amount of personal information about an account necessary to identify who the account belongs to. This is intended to be used in any and all APIs that expose accounts to admins or anyone other than the user themself who should not access their full profile. We should not be making a different version of this model for each similar API, but rather reusing this to ensure we are enforcing the same limitations on the data available. A better name for this model could be `AdminRestrictedPIIAccount` or something similar. If we wanted to expose a different subset of account data on a new API we should create a similar model with a meaningful and distinct name so we know when to use one vs. the other.

### Explanation
We realize that this understanding differs from yours in several key ways, but here is an explanation of why we still prefer this definition.
1. *Maintainability:** If we have a separate identical schema for each request/response that means that if we change a model we will need to go back and update every individual schema which will take more time and make it very difficult to ensure that all of the necessary schemas are updated. As the number of APIs grow it will be hard to remember every single individual API that uses a given model.
2. **Development speed:** Making separate schemas for each request/response will slow down development and require us to do additional work to create a new schema whenever we build a new API, even if an identical schema already exists. 
3. **Clarity:** While I do understand your point about using the models section to understand the internal Go structure of the model, I feel that this is not entirely necessary since these are the API docs and if a field is not exposed to the caller through an API then there is no reason we need them to be aware of these internal implementation details. In my view it is best to only present the information about the interface in these docs to keep them clear and focused.
4. **Consistency:** If we use a different schema for every request/response even if it is the same model being sent then my feeling is that the API docs will be more difficult to interpret. For example, the caller loses the ability to easily identify that it is an `Account` being sent to them and instead they see something with a non-meaningful name that just reflects the API they are calling. They will need to look through all of the individual fields to understand that the model being used between two APIs is actually the same model which will make it more difficult for them to know what models are necessary to build on their end.
5. **Readability:** If we have many duplicate request/response schemas, we will end up cluttering up the models section of our docs which will mean that readers have to look through a longer list to find the models they are looking for.

I realize that this is a larger conversation and that there is no definitive correct strategy to follow here, but hopefully this is helpful for you to understand the intention between the changes we made. Please let us know what you think of the organization we proposed here and the explanation we provided. Please also let us know If you have any questions or disagree with any of these points, or if we misunderstood anything about your explanation. 

Thanks again for your input!

shurwit commented:
Hi @petyos, thanks for the suggestion and clarification! 

As a side note we fixed the issue preventing us from upgrading to the latest version of `kin-openapi`.

> I agree it is a long conversation and there is no definitive correct strategy. As we are already in an advanced stage of development I would suggest to mix our understandings with minimal changes. I would suggest: Let's have 4 types schemas - 3 related with APIs and 1 for the Data model of the system: **APIs data model** This is general what meaning you currently give to the current model schemas.  So here we would have: AccountFileds.yaml - all simple fields of the Account entity(without composition with other entities) AdminRestrictedPIIAccount.yaml - some substructure of the Account entity etc **APIs Requests and Responses schemas:** Specific schemas for specific requests and responses **Data model schemas:** It represents the data model of the system and is a mirror of it.  It could Have AccountFields.yaml + all other entities compositions
> 
> ```
> Account
> {
> fields AccountFields.yaml
> 
> + all composition fields
> app_org ApplicationOrganization
> ...
> }
> ```
> 
> ///////////// What do you think about this? Thanks.

I think I understand what you are saying better now. I still don't fully see the need for the "Data model schemas" but I think the strategy you proposed to manage them is a good one. If we are going to continue supporting these I would like to propose a small modification to the structure you outlined to help keep things organized and make the purpose of each schema clear. Here is the file structure I would suggest for the `schemas/` directory:

```
fields/
  Account.yaml (name: AccountFields)
web/
  models/
    user/
      Account.yaml (name: Account)
  apis/ (matches the current docs/schemas/apis/)
internal/
  user/
      Account.yaml (name: InternalAccount)
```
* `fields`: Contains the root level primitive fields (no composition) for the model to be used in both `web/models/` and `internal/` schemas. It is also acceptable to reference these directly from the API definitions if they match the exact requirements.
* `web`: Contains all schemas that are used in APIs. Note that these schemas cannot contain recursive/circular references.
  * `models`: Contains all standard models for use in APIs and are composed of `fields` and any additional IDs or other properties using the [`allOf`](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/) keyword to merge and flatten the elements.
  * `apis`: Same as the existing request/response schemas. Contains schemas that are specific to an individual API request/response. Should only be used if no standard `web/model` fits. `shared` requests/responses should be used if the type is not a standard model, but is used by multiple APIs (eg. across services and admin endpoints). 
* `internal`: Schemas that reflect the internal structs for standard models composed of `fields` and other `internal` models using the [`allOf`](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/) keyword to merge and flatten elements. These are for informational purposes only and should not be used directly in API definitions. Because of this, these schemas may contain recursive/circular references, but should be ignored when generating Go types.

**Note:** No schemas should ever be exact duplicates.

shurwit commented:
Hi @petyos, I started attempting to implement the structure I described above and I am woried that it will still introduce a lot of extra work just to maintain the `internal` models since even with the shared `fields` we will need to essentially define every model twice and remember to update both when any changes occur. 

Could you please explain why you believe it is important to maintain the `internal` models even though they are not being used in the APIs and this information is not really needed by the caller? While it may not detract from the quality of the docs, it does make the process of defining and maintaining these docs significantly more complicated and time consuming. I am still having a hard time understanding what benefit these provide that make them worth the extra effort.

petyos commented:
> Hi @petyos, I started attempting to implement the structure I described above and I am woried that it will still introduce a lot of extra work just to maintain the `internal` models since even with the shared `fields` we will need to essentially define every model twice and remember to update both when any changes occur.
> 
> Could you please explain why you believe it is important to maintain the `internal` models even though they are not being used in the APIs and this information is not really needed by the caller? While it may not detract from the quality of the docs, it does make the process of defining and maintaining these docs significantly more complicated and time consuming. I am still having a hard time understanding what benefit these provide that make them worth the extra effort.

Hi @shurwit, my point of view was that we already have data model schemas which cover the data model schemas in the way I treat it at acceptable level of implementation. But indeed, introducing `internal` at this stage will trigger a lot of changes. Feel free to ignore the `Data model schemas/internal` as you do not see a reason to support it. It was an extra information.

petyos reviewed (dismissed):
LGTM

shurwit commented:
> Hi @shurwit, my point of view was that we already have data model schemas which cover the data model schemas in the way I treat it at acceptable level of implementation. But indeed, introducing `internal` at this stage will trigger a lot of changes. Feel free to ignore the `Data model schemas/internal` as you do not see a reason to support it. It was an extra information.

Hi @petyos, thanks for the explanation and for approving. I see what you mean about these data model schemas already existing. We were trying to simplify things and reduce the amount of documentation overhead going forward, and I do feel that maintaining them will complicate things. Hopefully you agree that this will simplify things. Thanks again!

petyos reviewed (dismissed):
LGTM

shurwit reviewed (dismissed):
Thanks @roberlander2, LGTM!
